### PR TITLE
Added Micro Speech support for ESP

### PR DESCRIPTION
**Pull request by AdityaHPatwardhan**
(Work In Progress)
This PR adds support for micro speech example on ESP. To execute the example you would require an ESP board with an inbuilt microphone  ( preferably ESP-EYE ) .( or you can connect mic externally to any esp board but that will require to make respective changes in audio_provider.cc)

- [x] To fix some build system issues
microfrontend/lib srcs and kissfft are added in CMakeLists.txt of main directory , they need to be removed from srcs makefile variable of main dir (  for now the esp/Makefile.inc adds these srcs in  makefile variable of component/tfmicro/CMakeLists.txt where they should be added ) 
- [x] Fix errors in headers for recognize_commands.h
a change of path in header micro_model_settings.h which causes error
- [x] A way to copy `sdkconfig.defaults’ and ‘Kconfig.projbuild’ ( espressif specific files) from `examples/micro_speech/esp/‘ to their respective location 
dest. for ‘sdkconfig.defaults’ -> ‘micro_speech/esp-idf’ ( the project folder)
dest. for ‘Kconfig.projbuild’ -> ‘micro_speech/esp-idf/main/‘ (`main` folder of the project). ( Note:  Kconfig.projbuild is not required for this project in particular, but it is required for [Person detection](https://github.com/tensorflow/tensorflow/pull/34751)

### CODE REVIEW

**Comment by mahavirj** [comment]
@fredrec

**Comment by rthadur** [comment]
@flyingraijin9 please sign CLA when you are done with changes

**Comment by AdityaHPatwardhan** [comment]
@fredrec please have a look at the build issues listed in PR description , as I believe you wrote the initial buid templates and helper_functions for esp  in tensorflow.

**Comment by AdityaHPatwardhan** [comment]
@googlebot I signed it!

**Comment by AdityaHPatwardhan** [comment]
@googlebot I signed it!

**Comment by fredrec** [comment]
Hi Aditya, thank you for this PR. 

About the build issue, I will send a patch that will allow to do this more cleanly.

In the end I could compile and flash to an ESP-EYE board, but unfortunately could not have it recognize voice commands. In particular, it never recognized `yes`. Did you have more success ?

**Comment by AdityaHPatwardhan** [comment]
@fredrec , Yes I had it correctly recognising my command most of the times , ( provided the environment was quiet one and I said yes,no in a specific tone ).

**Comment by AdityaHPatwardhan** [comment]
@fredrec @rthadur , PTAL , I have updated the `CMakeLists.txt.tpl` fixes from my side, it adds the `O3` flag which was not previously used while compiling , I have also uploaded a video for you to see at this [link](https://drive.google.com/file/d/1AYC7LsKUb1JOaByf1JbYwU1xwvUL7HDD/view) ,  Also Please provide the patch to fix the build issues and if there are no further issues remaining , and Please help merge this MR ASAP , Thank You

While testing  Please copy `sdkconfig.defaults` from `micro_speech/esp` into `esp-idf` folder and then build it ( provided any sdkconfig or sdkconfig.old in same folder is removed first )
_Note: While testing the example at start it takes roughly about 5-10 seconds to give first output , after that everything is instantaneous_

**Review by fredrec — COMMENTED** [review]
Thanks for your patience and sorry for the long time this PR got unattended.

After some minor modifications I could get the code to build and successfully flash an ESP-EYE board.
Here is the fix I had to apply: https://github.com/AdityaHPatwardhan/tensorflow/pull/1 Please have a look.

File: tensorflow/lite/micro/tools/make/templates/esp/components/tfmicro/CMakeLists.txt.tpl [hunk 1]
```diff
@@ -31,8 +31,7 @@ target_compile_options(${COMPONENT_LIB} PRIVATE
...
   -Wno-missing-field-initializers
   -Wno-pointer-sign
-  -Wno-type-limits)
+  -Wno-type-limits
+  %{CC_FLAGS}%
+  %{CXX_FLAGS}%)
 
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} %{CC_FLAGS}%")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} %{CXX_FLAGS}%")
-set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} %{LINKER_FLAGS}%")
```
[the remark this answers is not in the document]
Comment (AdityaHPatwardhan) [reply tensorflow/lite/micro/tools/make/templates/esp/components/tfmicro/CMakeLists.txt.tpl]: The thing is, with `set(CMAKE_C_FLAGS XXXX)` ,it doesn’t set the required flags in esp-idf , as we have a wrapper `idf.py` over `Cmake`. It over-rights them internally.So earlier the flags were not being applied. To add the CFLAGS, or CXXFLAGS , We need to add them with target_compile_options().

**Comment by AdityaHPatwardhan** [comment]
> Thanks for your patience and sorry for the long time this PR got unattended.
> 
> After some minor modifications I could get the code to build and successfully flash an ESP-EYE board.
> Here is the fix I had to apply: [AdityaHPatwardhan#1](https://github.com/AdityaHPatwardhan/tensorflow/pull/1) Please have a look.

Thank You @fredrec for the build fixes.I have cherry-picked your fixes in my PR, They work fine. 
I am afraid There is one more thing required here i.e. a way to copy `sdkconfig.defaults` from `examples/micro_speech/esp` to the home folder of the project (`micro_speech/esp-idf/` ). , I am sorry for not adding this in the checklist of the description above as I had added it in [person detection PR ](https://github.com/tensorflow/tensorflow/pull/34751) ( and I thought these issues would be fixed at same time.).

File: tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Adding some esp specific files in the main CMakeLists.txt
+ESP_MICRO_SPEECH_SRCS := \
+tensorflow/lite/micro/examples/micro_speech/esp/ringbuf.c \
```
Comment (fredrec) [inline tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc]: When defining variables across multiple lines, the last line should omit the trailing `\` (backslash).

File: tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+ESP_MICRO_SPEECH_HDRS := \
+tensorflow/lite/micro/examples/micro_speech/esp/ringbuf.h \
```
Comment (fredrec) [inline tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc]: ditto

File: tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+tensorflow/lite/experimental/microfrontend/lib/frontend_util.c \
+tensorflow/lite/experimental/microfrontend/lib/log_lut.c \
+tensorflow/lite/experimental/microfrontend/lib/log_scale.c \
+tensorflow/lite/experimental/microfrontend/lib/log_scale_util.c \
+tensorflow/lite/experimental/microfrontend/lib/noise_reduction.c \
+tensorflow/lite/experimental/microfrontend/lib/noise_reduction_util.c \
+tensorflow/lite/experimental/microfrontend/lib/pcan_gain_control.c \
+tensorflow/lite/experimental/microfrontend/lib/pcan_gain_control_util.c \
+tensorflow/lite/experimental/microfrontend/lib/window.c \
+tensorflow/lite/experimental/microfrontend/lib/window_util.c \
```
Comment (fredrec) [inline tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc]: ditto

File: tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc (hunk 2 above)
Comment (fredrec) [reply tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc]: Thanks for the update.

> I am afraid There is one more thing required here i.e. a way to copy `sdkconfig.defaults` from `examples/micro_speech/esp` to the home folder of the project (`micro_speech/esp-idf/` ).

I updated the PR with a new commit: https://github.com/AdityaHPatwardhan/tensorflow/pull/1/commits/3b429c79fb5862e51a26bd87ebc0d34c2815b3ba

Does this fix the issue ?

File: tensorflow/lite/micro/tools/make/templates/esp/components/tfmicro/CMakeLists.txt.tpl (hunk 1 above)
[the remark this answers is not in the document]
Comment (fredrec) [reply tensorflow/lite/micro/tools/make/templates/esp/components/tfmicro/CMakeLists.txt.tpl]: I see. And I guess it is better to ignore the linker flags for this platform ?

File: tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc (hunk 2 above)
Comment (fredrec) [reply tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc]: Oops, I did not notice the `Kconfig.projbuild` requirement. Let's handle that in the person detector PR instead then.

File: tensorflow/lite/micro/tools/make/templates/esp/components/tfmicro/CMakeLists.txt.tpl (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdityaHPatwardhan) [reply tensorflow/lite/micro/tools/make/templates/esp/components/tfmicro/CMakeLists.txt.tpl]: No, Thanks for noticing , I had made the necessary change some while ago ,along with addition of one CFLAG ,  I am yet to push all those changes .I will push those with fixing all other suggestions.

File: tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc (hunk 2 above)
Comment (AdityaHPatwardhan) [reply tensorflow/lite/micro/examples/micro_speech/esp/Makefile.inc]: > Does this fix the issue ?

Yes , Thanks for the build fixes , I have merged your PR into my branch.

**Comment by AdityaHPatwardhan** [comment]
@fredrec , I have made the necessary changes,The example now builds successfully without any build issue.
PTAL.

**Comment by fredrec** [comment]
Everything seems to be working fine.
@petewarden can you please merge this PR ?

**Comment by AdityaHPatwardhan** [comment]
Fixed some Compiler warnings , Example is working properly ( less surrounding noise is preferable)

**Review by fredrec — APPROVED** [review]

**Pull request merged**